dali/kernels/slice/slice_gpu.h

```diff
@@ -0,0 +1,134 @@
+#include <vector>
+#include "dali/core/common.h"
+#include "dali/core/convert.h"
+#include "dali/core/dev_array.h"
+#include "dali/core/error_handling.h"
+#include "dali/kernels/kernel.h"
+
+namespace dali {
+namespace kernels {
+
+template <std::size_t Dims>
+struct SliceArgsDev {
+  DeviceArray<int64_t, Dims> in_strides;
+  DeviceArray<int64_t, Dims> out_strides;
+  DeviceArray<int64_t, Dims> anchor;
+};
+
+template <typename OutputType, typename InputType, std::size_t Dims>
+__global__ void SliceKernel(OutputType *__restrict__ output,
+                            const InputType *__restrict__ input,
+                            SliceArgsDev<Dims> slice_args_dev,
+                            unsigned int total_pixels) {
+  unsigned int idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (idx >= total_pixels) {
+    return;
+  }
+
+  unsigned int out_idx = idx;
+  unsigned int in_idx = 0;
+  for (std::size_t d = 0; d < Dims; d++) {
+    unsigned int i_d = idx / slice_args_dev.out_strides[d];
+    idx = idx % slice_args_dev.out_strides[d];
+    in_idx += (slice_args_dev.anchor[d] + i_d) * slice_args_dev.in_strides[d];
+  }
+  output[out_idx] = clamp<OutputType>(input[in_idx]);
+}
+
+template <std::size_t Dims>
+struct SliceArgs {
+  std::array<int64_t, Dims> anchor;
+  std::array<int64_t, Dims> shape;
+};
+
+template <std::size_t Dims, typename Shape>
+std::array<int64_t, Dims> GetStrides(const Shape& shape) {
+  std::array<int64_t, Dims> strides;
+  strides[Dims - 1] = 1;
+  for (std::size_t d = Dims - 1; d > 0; d--) {
+    strides[d - 1] = strides[d] * shape[d];
+  }
+  return strides;
+}
+
+template <typename InputType, typename OutputType, std::size_t Dims>
+class DLL_PUBLIC SliceGPU {
+ public:
+  SliceGPU() = default;
+
+  DLL_PUBLIC KernelRequirements Setup(KernelContext &context,
+                                      const InListGPU<InputType, Dims> &in,
+                                      const std::vector<SliceArgs<Dims>> &slice_args) {
+    KernelRequirements req;
+
+    DALI_ENFORCE(slice_args.size() == static_cast<std::size_t>(in.size()),
+      "Number of samples and size of slice arguments should match");
+
+    std::vector<TensorShape<Dims>> output_shapes;
+    for (int i = 0; i < in.size(); i++) {
+      auto in_sample_shape = in.tensor_shape(i);
+      TensorShape<Dims> out_sample_shape(slice_args[i].shape);
+      auto &anchor = slice_args[i].anchor;
+
+      for (std::size_t d = 0; d < Dims; d++) {
+        DALI_ENFORCE(anchor[d] >= 0 && (anchor[d] + out_sample_shape[d]) <= in_sample_shape[d],
+          "Slice " + std::to_string(i) + " dimension " + std::to_string(d) +
+          " is out of bounds : " + "anchor[" + std::to_string(anchor[d]) +
+          "] size[" + std::to_string(out_sample_shape[d]) + "] input dimension size[" +
+          std::to_string(in_sample_shape[d]) + "]");
+      }
+
+      output_shapes.push_back(out_sample_shape);
+    }
+    req.output_shapes.push_back(TensorListShape<Dims>{output_shapes});
```

Comment:
...and finally `push_back(std::move(output_shape))`

Reply:
done